middleware/src/middleware.go

```diff
@@ -522,37 +668,100 @@ func (middleware *Middleware) SetRootPassword(args rpcmessages.SetRootPasswordAr
+				rpcmessages.ErrorSetNeedsTwoArguments,
+			})
+
+			return rpcmessages.ErrorResponse{
+				Success: false,
+				Message: strings.Join(out, "\n"),
+				Code:    errorCode,
+			}
 		}
+
 		return rpcmessages.ErrorResponse{Success: true}
 	}
-	return rpcmessages.ErrorResponse{Success: false, Message: "invalid password"}
+
+	return rpcmessages.ErrorResponse{
+		Success: false,
+		Message: "The password has to be at least 8 chars. An unicode char is counted as one.",
+		Code:    rpcmessages.ErrorSetRootPasswordToShort,
+	}
 }
 
 // runBBBCmdScript runs the bbb-cmd.sh script.
 // The script executes commands like for example mounting a USB drive, doing a backup and copying files.
-func (middleware *Middleware) runBBBCmdScript(args []string) (out []byte, err error) {
-	script := middleware.environment.GetBBBCmdScript()
-	cmdAsString := script + " " + strings.Join(args, " ")
-	out, err = exec.Command(script, args...).Output()
-	if err != nil {
-		// no error handling here, only logging.
-		log.Printf("Error: The command '%s' exited with the output '%v' and error '%s'.\n", cmdAsString, string(out), err.Error())
-	}
+func (middleware *Middleware) runBBBCmdScript(args []string) (outputLines []string, err error) {
+	outputLines, err = runCommand(middleware.environment.GetBBBCmdScript(), args)
 	return
 }
 
 // runBBBConfigScript runs the bbb-config.sh script.
 // The script changes the system configuration in redis by setting or unsetting the appropriate keys.
 // If necessary the affected services are restarted.
-func (middleware *Middleware) runBBBConfigScript(args []string) (out []byte, err error) {
-	script := middleware.environment.GetBBBConfigScript()
-	cmdAsString := script + " " + strings.Join(args, " ")
-	out, err = exec.Command(script, args...).Output()
+func (middleware *Middleware) runBBBConfigScript(args []string) (outputLines []string, err error) {
+	outputLines, err = runCommand(middleware.environment.GetBBBConfigScript(), args)
+	return
+}
+
+// runCommand runs the passed command and returns the combined stdout and stderr output.
+// If the command could not be run, err is not nil.
+func runCommand(command string, args []string) (combinedLines []string, err error) {
+	cmd := exec.Command(command, args...)
+	log.Printf("executing command: %s %s", command, strings.Join(args, " "))
+
+	rawstdoutStderr, err := cmd.CombinedOutput()
 	if err != nil {
-		// no error handling here, only logging.
-		log.Printf("Error: The command '%s' exited with the output '%v' and error '%s'.\n", cmdAsString, string(out), err.Error())
+		// no error handling here, just logging
+		log.Printf("Error executing command '%s %s': '%s'", command, strings.Join(args, " "), err.Error())
 	}
-	return
+
+	combined := strings.TrimSuffix(string(rawstdoutStderr), "\n")
+	combinedLines = strings.Split(combined, "\n")
+	return combinedLines, err
+}
+
+func handleBBBScriptError(outputLines []string, err error, possibleErrors []rpcmessages.ErrorCode) rpcmessages.ErrorCode {
```

Comment:
I'm not sure about the `possibleErrors []rpcmessages.ErrorCode` parameter. 
This means that anyone implementing new RPC calls will have to know about all the possible errors and pass the appropriate ones to `handleBBBScriptError`m this seems like a lot of overhead for anyone not familiar with the errors, and a lot of manual checking to make sure you included all the necessary errors, which might also change if we add more errors/codes in the future

Reply:
I see your point, but I think somebody adding an RPC that executes a script should need to know what the script does and which errors it can throw (either by checking or having a spec from the script author). It's currently designed that `possibleErrors` only needs to take the exact error that the  _method_ invoked in the script can throw. All other possible errors like e.g. `ErrorScriptNotSuperuser` are checked by `handleBBBScriptError()`. The `possibleErrors` approach has the benefit that if the script or the code don't match an `ErrorUnhandled` ("UNKNOWN") is returned. That should be something never occurring besides when there is an implementation error in either the the RPC or the scripts. 

(I'll add some of this as a comment though)

Reply:
I would still really disagree on this. Passing possible error codes to a function just adds an extra layer of indirection. If there is a function such as `handleBBBScriptError()` it should handle all errors, at least that's what I would expect as the caller. If it accepts possible errors as arguments, you might as well handle those possible errors before even calling `handleBBBScriptError()` since you already know what those errors are. Still not ideal though. I actually think this relates nicely to Kaspar's design by contract. After receiving an error from the base and calling `handleBBBScriptError()` I would expect it to return the appropriate error code. Having the additional responsibility as the caller to pre-define some of the function's possible return values just feels really really strange.

Reply:
What do you think about having the error codes in a `map[string]ErrorCode` and doing lookups in `handleBBBScriptErrorCode()` ? _(I not sure if a  `map[string]ErrorCode` would be the best datatype for that. We probably don't want to add every new `ErrorCode` to it manually ... Will need to think about that a bit more.)_ The downside is that we wouldn't know when the script returns a not valid error code for that script call (e.g. `flashdrive check` returns `SET_BITCOINETWORK_INVALID_VALUE` and we happily pass that back as a the RPC ErrorCode).

Reply:
Can you give an example for when/why a script would return an invalid error code for that script call? Then it's the responsibility of the script to return the appropriate code and the middleware shouldn't be sanitizing what the scripts return. Why would for example `flashdrive check` return `SET_BITCOINETWORK_INVALID_VALUE`? That seems to defeat the purpose of returning appropriate error codes from the scripts if you can't rely on them and just do separate lookups/sanity checks in the middleware again. 

I suppose my ideal scenario is that since the scripts should return informative, unique and appropriate error codes that the middleware can just pass them on to the app which can then handle the lookup since it needs to do translation for each possible error anyway, otherwise give an unknown error if it doesn't know how to handle it. But maybe I'm missing something why this is not possible/good solution.

Reply:
> I would expect it to return the appropriate error code. Having the additional responsibility as the caller to pre-define some of the function's possible return values just feels really really strange.

But it's not a "caller", it's all in the same package/file, so it's an implementation detail where you define the valid response codes per api call.

It does make sense to specify clearly which errors an api endpoint can return, and not accidentally "leak" more error codes as you introduce them in other parts of the code.

Imho "Unhandled" is just misleading naming, better would be "Other" or "Unexpected".

Reply:
> But it's not a "caller", it's all in the same package/file, so it's an implementation detail where you define the valid response codes per api call.
> 
> It does make sense to specify clearly which errors an api endpoint can return, and not accidentally "leak" more error codes as you introduce them in other parts of the code.

Yes, agreed that specifying the errors makes sense, I just wouldn't do it in explicitly in every `handleBBBScriptError()` call. An external lookup table, something like the `map[string]ErrorCode` map Timo is proposing would make more sense I think. That way when you call `handleBBBScriptError()` the implementation can already do all the error code handling.

Reply:
Offline decision: leave as is and iterate on in the future if needed.